rfcs/convergence/input-native-element-props.md

```diff
@@ -0,0 +1,195 @@
+
+#### [React Final Form](https://github.com/final-form/react-final-form)
+
+## Detailed Design or Proposal
+
+We don't have one suggested/preferred solution yet, because all of them have some significant problems.
+
+The examples below assume you're rendering a Checkbox and this is roughly your desired HTML output (it would also include a generated ID to associate the label and input):
+
+```html
+<div>
+  <input type="checkbox" name="foo" checked />
+  <label>sample</label>
+</div>
+```
+
+### Option 1: All native props always applied to root (current behavior)
+
+Given this:
+
+```tsx
+<Checkbox name="foo" checked>
+  sample
+</Checkbox>
+```
+
+You get roughly this HTML, which is not useful:
+
+```html
+<div name="foo" checked>
+  <input type="checkbox" />
+  <label>sample</label>
+</div>
+```
+
+So to specify a value, you'd have to use slot props:
+
+```tsx
+<Checkbox input={{ name: 'foo', checked: true }}>sample</Checkbox>
+```
+
+#### Pros
+
+- Complete consistency between all components
+
+#### Cons
+
+- Very counterintuitive API
+  - We could use typings to ensure that doing e.g. `<Checkbox checked/>` caused a TS error, but it still seems very unnatural
+- Might not work as easily with 3rd-party form libraries? (needs more research)
+
+### Option 2: Most native props applied to root, selected props applied to "actual" element
+
+Given this, and an implementation which uses a list of special props (including `checked` and `name`) that are passed to the "actual" element:
+
+```tsx
+<Checkbox name="foo" checked>
+  sample
+</Checkbox>
+```
+
+You'd get the desired HTML:
+
+```html
+<div>
+  <input type="checkbox" checked name="foo" />
+  <label>sample</label>
+</div>
+```
+
+The problem is, it's unclear where some of the other props ought to be applied, such as `id`, `className`, and event handlers.
+
+#### Pros
+
+- Better API for input components (and maybe other special-case components)
+
+#### Cons
+
+- Not clear how to determine which top-level props should be passed down to the `input`
+  - If `checked` and `name` (and a few others) go to the `input`, the user might expect this behavior for all props and then be surprised
+- Not clear what happens if a prop may be needed in multiple places
+  - Example: if top-level `className` or `id` is passed to the `input`, what happens if they also want to give the root a `className` or `id`?
+- Inconsistent between components (reduces API clarity)
+
+### Option 3
+
+Let the native input element be the "root" slot, but use a `wrapper` slot as the actual root DOM element. Here's an example:
+
+```tsx
+<slots.wrapper>
+    <div {...checkmark} />
+    <slots.root {...slotProps.root} /> <!-- input element -->
+    <slots.label {...slotProps.label} />
+</slots.wrapper>
+```
+
+So doing this would give the desired HTML:
+
+```html
+<div>
+  <div />
+  <!-- checkmark -->
```

Comment:
Probably best to remove this checkmark HTML (also in other examples) since it's a bit distracting from the main point and prettier reformatted it in an unhelpful way